cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/PullStrategy.java

```diff
@@ -262,6 +262,13 @@ private void replicate()
             if (changeFeeds.size() > 0) {
                 batchChangesProcessed = processOneChangesBatch(changeFeeds);
                 state.documentCounter += batchChangesProcessed;
+            } else {
```

Comment:
I'm a little concerned about this being in an `else` block without additional checks. Whilst this works for your golden path case of having some new changes (all filtered) and thus seeing 0 changes, but having a new `lastSeq` I think there are other cases where it is going to force the writing of a checkpoint even though there has been no change to the `lastSeq`.
For example, in cases of 0 changes received (with no filtering) or an integer multiple of `changeLimitPerBatch` changes this else will cause writing the same checkpoint/lastSeq we already know once more again before breaking out of the loop.

I don't think this concern is definitely a blocker for this change, but I think we ought to at least consider the potential impact of writing those redundant checkpoints vs the complexity of moving the retrieval of the current known `lastSeq` out of `processOneChangesBatch` so that we could validate if it changed before writing the checkpoint in something like an `else if (!knownLastSeq.equals(changeFeeds.getLastSeq())) {...}`.

Reply:
Thanks @ricellis for your review. 

I've reworked this PR including your proposal. The current checkpoint is retrieved on the main loop and setted back only it changed. I agree with you that it is better to protect the checkpoint from redundant updates. 

I also added some tests related to checkpoint recording. I didn't find many tests related to checkpoints.

Reply:
Thanks, this looks really good.
I'm running our on-device tests then hopefully we'll be able to proceed.